src/main/scala/com/microsoft/hyperspace/Hyperspace.scala

```diff
@@ -165,19 +171,42 @@ class Hyperspace(spark: SparkSession) {
   }
 }
 
-object Hyperspace {
-  private lazy val context = new ThreadLocal[HyperspaceContext]
+object Hyperspace extends Logging {
+  // Maintain HyperspaceContext for each Spark session, since Hyperspace depends on
+  // the session's properties such as configs, etc.
+  // Value as (thread id, hyperspace context) to enforce one client thread per Spark session
+  // as Hyperspace doesn't guarantee thread safety.
+  private lazy val contexts =
+    new ConcurrentHashMap[SparkSession, (Long, HyperspaceContext)]().asScala
 
   private[hyperspace] def getContext(spark: SparkSession): HyperspaceContext = {
-    val ctx = context.get()
-    if (ctx == null || !ctx.spark.equals(spark)) {
-      // Ensure the Spark session the caller provides is the same as
-      // the one HyperspaceContext is using because Hyperspace depends on the
-      // session's properties such as configs, etc.
-      context.set(new HyperspaceContext(spark))
+    val ctx = contexts.get(spark)
+    val threadId = Thread.currentThread().getId
+    if (!ctx.isDefined) {
+      contexts.synchronized {
+        if (contexts.size >= HyperspaceConf.maxNumAllowedConcurrentSparkSessions(spark)) {
+          throw HyperspaceException(
+            "Too many Spark sessions are using Hyperspace. Please increase the config `" +
+              IndexConstants.MAX_NUM_ALLOWED_CONCURRENT_SPARK_SESSIONS + "`. " +
+              s"Current limit: ${contexts.size}")
+        }
+        val newCtx = new HyperspaceContext(spark)
+        contexts.put(spark, (threadId, newCtx))
```

Comment:
Should we check the return value of `put` and throw an exception if it returns `Some`? There could be multiple threads sharing the same spark session that hits `if (!ctx.isDefined)` at the same time.